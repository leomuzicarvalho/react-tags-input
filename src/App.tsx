import React, { useState } from 'react'
import TagsInput from './TagsInput'

export interface deleteOptions {
  item?: string
  deleteLast?: boolean
}

const App: React.FC = () => {
  const [tags, setTags] = useState<string[]>([
    'contato@rarolabs.com.br',
    'nao-responda@rarolabs.com.br',
  ])

  const validateEmail = (email: string) => {
    const re =
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
    return re.test(email.toLowerCase())
  }

  const handleSelecetedTags = (items: string | string[]) => {
    if (items instanceof Array) {
      items = items.filter((it: string) => validateEmail(it) && !tags.includes(it))
      setTags([...tags, ...items])
      return
    }
    if (validateEmail(items) && !tags.includes(items)) setTags([...tags, items])
  }

  const handleDeleteTag = (options: deleteOptions) => {
    const { deleteLast, item } = options
    if (deleteLast) {
      setTags((oldItems) => oldItems.slice(0, -1))
      return
    }
    setTags((oldItems) => oldItems.filter((it: string) => it !== item))
  }

  return (
    <div className="App">
      <TagsInput
        onDeleteTag={(options) => () => handleDeleteTag(options)}
        onSelectTags={handleSelecetedTags}
        fullWidth
        variant="outlined"
        id="tags"
        name="tags"
        placeholder="add Tags"
        label="tags"
        tags={tags}
      />
    </div>
  )
}

export default App
