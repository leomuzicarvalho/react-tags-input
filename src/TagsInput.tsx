import React from 'react'
import Chip from '@material-ui/core/Chip'
import { TextField } from '@material-ui/core'
import { withStyles } from '@material-ui/core/styles'
import { OutlinedTextFieldProps } from '@material-ui/core/TextField'

import { deleteOptions } from '@/App'

interface TagsInputProps extends OutlinedTextFieldProps {
  onSelectTags: (items: string[] | string) => void
  onDeleteTag: (options: deleteOptions) => () => void
  tags: string[]
}

const CssTextField = withStyles({
  root: {
    '& .MuiInputBase-root': {
      display: 'flex',
      flexWrap: 'wrap',
    },
  },
})(TextField)

export default function TagsInput(props: TagsInputProps) {
  const { onSelectTags, onDeleteTag, tags, ...other } = props
  const [value, setValue] = React.useState('')

  return (
    <CssTextField
      onKeyDown={(e) => {
        if (e.key === 'Backspace' && tags.length > 0 && value === '') {
          e.preventDefault()
          onDeleteTag({ deleteLast: true })()
          return
        }
        if (e.key === 'Tab' || e.key === 'Enter') {
          e.preventDefault()
          if (!value.includes(',') && !value.includes(';')) {
            onSelectTags(value)
            setValue('')
            return
          }
          const aux = value.split(/(?:,|;)+/).filter((x) => x)
          const values = aux.length > 1 ? aux : aux[0]
          if (!values) {
            setValue('')
            return
          }
          onSelectTags(values)
          setValue('')
        }
      }}
      onChange={(e) => setValue(e.target.value)}
      value={value}
      InputProps={{
        startAdornment: tags.map((item, idx) => (
          <Chip
            key={`${item}-${idx}`}
            tabIndex={-1}
            label={item}
            onDelete={onDeleteTag({ item })}
            style={{ margin: '10px 2px' }}
          />
        )),
      }}
      {...other}
    />
  )
}
TagsInput.defaultProps = {
  tags: [],
}
