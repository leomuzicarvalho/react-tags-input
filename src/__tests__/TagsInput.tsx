import React from 'react'
import App from '@/App'
import TagsInput from '@/TagsInput'
import { cleanup, fireEvent, render } from '@testing-library/react'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  it('cria o snapshot do component', () => {
    const container = render(
      <TagsInput
        variant="outlined"
        id="tags"
        name="tags"
        placeholder="add Tags"
        label="tags"
        onSelectTags={() => void 0}
        onDeleteTag={() => () => void 0}
      />
    )
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('deve renderizar as tags enviadas por atributos', () => {
    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']

    const { debug } = render(
      <TagsInput
        variant="outlined"
        id="tags"
        name="tags"
        placeholder="add Tags"
        label="tags"
        onSelectTags={() => void 0}
        onDeleteTag={() => () => void 0}
        tags={emails}
      />
    )

    debug()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', () => {
    const component = render(<App />)
    const inputEl = component.getByPlaceholderText('add Tags')
    fireEvent.change(inputEl, {
      target: { value: 'asd@asd.com' },
    })
    fireEvent.keyDown(inputEl, {
      key: 'Enter',
    })

    expect(component.baseElement).toContainHTML('<span class="MuiChip-label">asd@asd.com</span>')
  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
    const component = render(<App />)
    const inputEl = component.getByPlaceholderText('add Tags')
    fireEvent.change(inputEl, {
      target: { value: 'asd@asd.com' },
    })
    fireEvent.keyDown(inputEl, {
      key: 'Tab',
    })

    expect(component.baseElement).toContainHTML('<span class="MuiChip-label">asd@asd.com</span>')
  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {
    const component = render(<App />)
    const inputEl = component.getByPlaceholderText('add Tags')
    fireEvent.keyDown(inputEl, {
      key: 'Backspace',
    })

    expect(component.queryByText('nao-responda@rarolabs.com.br')).toBeNull()
  })
})
