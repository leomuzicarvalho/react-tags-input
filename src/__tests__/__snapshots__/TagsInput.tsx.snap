// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
      data-shrink="true"
      for="tags"
    >
      tags
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
    >
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-3"
          style="width: 0px;"
        >
          <span>
            â€‹
          </span>
        </legend>
      </fieldset>
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
        id="tags"
        name="tags"
        placeholder="add Tags"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
